performance/serving-optimization.mdx

```diff
@@ -0,0 +1,74 @@
+
+## SQL Optimizations
+
+### When to create indexes
+
+Indexes in RisingWave are used to accelerate batch queries. See the basics of indexes in [CREATE INDEX](/sql/commands/sql-create-index).
+
+We determine how the index should be created by checking:
+
+1.  Which columns of the materialized views are used in the `SELECT` statement? These columns should all appear in the `INCLUDE` clause when creating the index.
+2.  Which columns of the materialized views are included in an optional `WHERE` condition in the batch queries? Suppose the batch query filters a column named timestamp with the condition `timestamp between t1 and t2`, then the column `timestamp` should be included in the `index_column`. The same principle applies to any other filter conditions such as equality and inequalities.
+
+We remark that the incremental maintenance of indexes in RisingWave is similar to the incremental maintenance of materialized views but with minimal computation. Therefore, it is cost-effective to create indexes in RisingWave. We encourage users to detect the patterns in batch queries and create indexes if the pattern occurs frequently or/and a batch query is slow.
+
+Examples:
+
+```sql
+CREATE TABLE t(k1 INT, k2 INT, v1 INT, v2 INT, PRIMARY KEY(k1,k2));
+
+--- idx_t_k2_partial_columns doesn't include column v2.
+CREATE INDEX idx_t_k2_partial_columns ON t(k2) INCLUDE (k1,v1);
+
+--- idx_t_k2_partial_columns is not utilized.
+EXPLAIN SELECT v2 FROM t WHERE k1=1;
+ BatchExchange { order: [], dist: Single }
+ └─BatchScan { table: t, columns: [v2], scan_ranges: [k1 = Int32(1)] }
+
+--- idx_t_k2_partial_columns is utilized.
+EXPLAIN SELECT v1 FROM t WHERE k2=1;
+ BatchExchange { order: [], dist: Single }
+ └─BatchScan { table: idx_t_k2_partial_columns, columns: [v1], scan_ranges: [k2 = Int32(1)] }
+
+--- idx_t_k2_partial_columns is utilized. However since it doesn't include column v2, the plan requires an additional lookup join.
+EXPLAIN SELECT v2 FROM t WHERE k2=1;
+ BatchLookupJoin { type: Inner, predicate: idx_t_k2_partial_columns.k1 IS NOT DISTINCT FROM t.k1 AND idx_t_k2_partial_columns.k2 IS NOT DISTINCT FROM t.k2 AND (t.k2 = 1:Int32), lookup table: t }
+ └─BatchExchange { order: [], dist: Single }
+   └─BatchScan { table: idx_t_k2_partial_columns, columns: [k2, k1], scan_ranges: [k2 = Int32(1)] }
+
+--- idx_t_k2_all_columns includes all columns.
+CREATE INDEX idx_t_k2_all_columns ON t(k2);
+
+--- idx_t_k2_all_columns is utilized. Compare the plan with the one that uses idx_t_k2_partial_columns.
+EXPLAIN SELECT v2 FROM t WHERE k2=1;
+ BatchExchange { order: [], dist: Single }
+ └─BatchScan { table: idx_t_k2_all_columns, columns: [v2], scan_ranges: [k2 = Int32(1)] }
+```
+
+### Leverage predicate pushdown
+
+Predicate pushdown allows filtering operations (predicates) to be applied as early as possible in the query pipeline. This means that instead of retrieving all data and then filtering it, RisingWave filters the data at the storage level.
+
```

Comment:
I think this section can be better called "Point get and range scan (by prefix of primary key)"?

BTW, this should be put above index section:
- index also uses this optimization to be efficient
- need to create index only when query pattern is not on PK

Reply:
> Point get and range scan (by prefix of primary key)

Doesn't seem very understandable to users as a title. In contrast, predicate pushdown is a known optimization concept in SQL.

Reply:
> BTW, this should be put above index section:

    index also uses this optimization to be efficient
    need to create index only when query pattern is not on PK

Done https://github.com/risingwavelabs/risingwave-docs/pull/401/commits/f931a7be762bb3a1d1c756b12a78279e14317e1e